### PR TITLE
[CPT-632] Avalanche Bridge Support

### PR DESCRIPTION
**Pull request by learyce**
Adding support for Rosetta to call unwrap on a configured list of contracts

https://github.com/ava-labs/avalanche-rosetta/pull/101 has been rebased on top of this branch

### CODE REVIEW

File: go.sum [hunk 1]
```diff
@@ -15,6 +15,20 @@ cloud.google.com/go v0.56.0/go.mod h1:jr7tqZxxKOVYizybht9+26Z/gUq7tiRzu+ACVAMbKV
 cloud.google.com/go v0.57.0/go.mod h1:oXiQ6Rzq3RAkkY7N6t3TcE6jE+CIBBbA36lwQ1JyzZs=
 cloud.google.com/go v0.62.0/go.mod h1:jmCYTdRCQuc1PHIIJ/maLInMho30T/Y0M4hTdTShOYc=
 cloud.google.com/go v0.65.0/go.mod h1:O5N8zS7uWy9vkA9vayVHs65eM1ubvY4h553ofrNHObY=
+cloud.google.com/go v0.72.0/go.mod h1:M+5Vjvlc2wnp6tjzE102Dw08nGShTscUx2nZMufOKPI=
```
Comment (dhrubabasu) [inline go.sum:18]: `go mod tidy -compat=1.17` should revert these changes

File: service/types.go [hunk 2]
```diff
@@ -32,14 +33,15 @@ type optionsWire struct {
 	GasLimit               string          `json:"gas_limit,omitempty"`
 	Nonce                  string          `json:"nonce,omitempty"`
 	Currency               *types.Currency `json:"currency,omitempty"`
+	UnwrapBridgeTx         bool            `json:"bridge_unwrap"`
 }
 
 func (o *options) MarshalJSON() ([]byte, error) {
 	ow := &optionsWire{
 		From:                   o.From,
-		To:                     o.To,
```
Comment (dhrubabasu) [inline service/types.go:40]: Why are we removing this?

Comment (dhrubabasu) [reply service/types.go:40]: Should be removed from the types if deliberate

File: Makefile [hunk 3]
```diff
@@ -99,10 +99,14 @@ check-testnet-data:
 check-testnet-construction:
 	rosetta-cli check:construction --configuration-file=rosetta-cli-conf/testnet/config.json
 
-# Perform the Testnet construction check for ERC-20s
+# Perform the Testnet construction check for ERC-20s transfers
```
Comment (dhrubabasu) [inline Makefile:102]: nit:
```suggestion
# Perform the Testnet construction check for ERC-20 transfers
```

File: cmd/server/config.go [hunk 4]
```diff
@@ -89,6 +91,19 @@ func (c *config) Validate() error {
 		}
 	}
 
+	if len(c.BridgeTokenList) != 0 {
+		for _, token := range c.BridgeTokenList {
```
Comment (dhrubabasu) [inline cmd/server/config.go:95]: Should we verify that the token address is a bridge token?

File: service/service_construction.go [hunk 5]
```diff
@@ -786,8 +1092,35 @@ func parseErc20TransferData(data []byte) (*ethcommon.Address, *big.Int, error) {
...
+	chainID := new(big.Int).SetBytes(data[37:])
+
+	if chainID.Uint64() != 0 {
+		return nil, nil, fmt.Errorf("incorrect chainId value")
+	}
+
+	return amount, chainID, nil
+}
+
 func getTransferMethodID() []byte {
```
Comment (dhrubabasu) [inline service/service_construction.go:1114]: I think a `getMethodID` function that takes in a fn signature would be good here

Comment (dhrubabasu) [reply service/service_construction.go:1114]: #101

File: cmd/server/config.go (hunk 4 above)
Comment (learyce) [reply cmd/server/config.go:95]: There's not really a good check for what is/isn't a bridge token

File: service/service_construction.go [hunk 6]
```diff
@@ -23,9 +23,12 @@ import (
 
 const (
 	genericTransferBytesLength = 68
+	genericUnwrapBytesLength   = 68
```
Comment (patrick-ogrady) [inline service/service_construction.go:26]: Is this referenced somewhere? May be good practice to indicate how these are computed.

File: rosetta-cli-conf/testnet/avalanche_unwrap.ros [hunk 7]
```diff
@@ -0,0 +1,104 @@
...
+      "create_probability": 50
+    });
+    transfer.preprocess_metadata = {
+        "bridge_unwrap": true
+    };
+    transfer.confirmation_depth = "1";
+    transfer.operations = [
+      {
+        "operation_identifier":{"index":0},
+        "type":"ERC20_BURN",
```
Comment (patrick-ogrady) [inline rosetta-cli-conf/testnet/avalanche_unwrap.ros:95]: Can we make this `ERC20_UNWRAP`?

File: rosetta-cli-conf/testnet/avalanche_unwrap.ros [hunk 8]
```diff
@@ -0,0 +1,104 @@
...
+      "not_account_identifier":[{{sender.account_identifier}}],
+      "minimum_balance":{
+        "value": "0",
+        "currency": {{currency}}
+      },
+      "create_limit": 100,
+      "create_probability": 50
+    });
+    transfer.preprocess_metadata = {
+        "bridge_unwrap": true
```
Comment (patrick-ogrady) [inline rosetta-cli-conf/testnet/avalanche_unwrap.ros:89]: We can get rid of this once `ERC20_UNWRAP` is in place.

File: rosetta-cli-conf/testnet/avalanche_unwrap.ros (hunk 7 above)
Comment (patrick-ogrady) [reply rosetta-cli-conf/testnet/avalanche_unwrap.ros:95]: There appears to be an `Unwrap` event, so this should be possible. I see the potential difficulty in separating from `Burn` (which will be invoked during any `Unwrap`). 

However, I think this will be really hard for anyone to parse in isolation.

File: service/service_construction.go [hunk 9]
```diff
@@ -113,9 +120,10 @@ func (s ConstructionService) ConstructionMetadata(
 	}
 
 	metadata := &metadata{
-		Nonce:    nonce,
-		GasPrice: gasPrice,
-		GasLimit: gasLimit,
+		Nonce:          nonce,
+		GasPrice:       gasPrice,
+		GasLimit:       gasLimit,
+		UnwrapBridgeTx: input.UnwrapBridgeTx,
```
Comment (patrick-ogrady) [inline service/service_construction.go:126]: We should really try to avoid specifying the tx type by providing a metadata arg. This conflicts with some of the Rosetta principles.

File: cmd/server/config.go (hunk 4 above)
Comment (dhrubabasu) [reply cmd/server/config.go:95]: I wonder if it makes sense to hardcode the bridge tokens in rosetta cc @patrick-ogrady

File: rosetta-cli-conf/testnet/avalanche_unwrap.ros (hunk 7 above)
Comment (rajranjan0608) [reply rosetta-cli-conf/testnet/avalanche_unwrap.ros:95]: Decided to go with the `ERC20_BURN` opcode.

File: service/service_construction.go (hunk 6 above)
Comment (rajranjan0608) [reply service/service_construction.go:27]: Added comment to explain these:

// 68 Bytes = methodID (4 Bytes) + param 1 (32 Bytes) + param 2 (32 Bytes)

**Comment by andrewkmin** [comment]
How will `unwrap` be parsed on the syncing side of things? Given that each unwrap has a `Transfer`, I'm guessing that'll suffice?

Example tx https://testnet.snowtrace.io/tx/0xc56da1523ae76a400999abce14b2bec06cd5a9adf2ba37bd577df4a3ab92ec28#eventlog

**Comment by rajranjan0608** [comment]
> How will `unwrap` be parsed on the syncing side of things?...

It will be parsed as erc20 burn op (`logs.topics.toAddress == 0x0`) as shown [here](https://github.com/ava-labs/avalanche-rosetta/blob/2b309ff13b198940e0f62bea839d9ab1476f806d/mapper/transaction.go#L462).

**Review by rajranjan0608 — APPROVED** [review]

**Comment by andrewkmin** [comment]
hey there 👋 we'd love to pick this back up again. any chance we can get it rebased? thank you 🙏

**Comment by andrewkmin** [comment]
hey team, where were we in terms of potentially discussing rosetta supporting arbitrary contract calls? there might come a time when we (or other clients) could want to do more than just unwrapping
